### PR TITLE
[Merged by Bors] - feat(FieldTheory/Perfect): add `rootsExpand[Pow]EquivRoots[_apply]`

### PR DESCRIPTION
**Pull request by acmepjz**
... which state that over a perfect integral domain `R` of characteristic `p`, `x ↦ x ^ p ^ n` is a bijection from the set of roots of `Polynomial.expand R (p ^ n) f` to the set of roots of `f`.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/FieldTheory/Perfect.lean [hunk 1]
```diff
@@ -111,14 +111,50 @@ theorem not_irreducible_expand (f : R[X]) : ¬ Irreducible (expand R p f) := by
...
+
+/-- If `f` is a polynomial over a perfect integral domain `R` of characteristic `p`, then there is
+a bijection from the set of roots of `Polynomial.expand R p f` to the set of roots of `f`.
+It's given by `x ↦ x ^ p`, see `rootsExpandEquivRoots_apply`. -/
+noncomputable def rootsExpandEquivRoots : (expand R p f).roots.toFinset ≃ f.roots.toFinset :=
+  ((frobeniusEquiv R p).image _).trans <| Equiv.Set.ofEq <| show _ '' (setOf _) = setOf _ by
+    ext r; obtain ⟨r, rfl⟩ := surjective_frobenius R p r
+    simp [expand_eq_zero (Fact.out : p.Prime).pos, (frobenius_inj R p).eq_iff, ← frobenius_def]
+
+theorem rootsExpandEquivRoots_apply (x) : (rootsExpandEquivRoots p f x).1 = x.1 ^ p := rfl
```
Comment (erdOne) [inline Mathlib/FieldTheory/Perfect.lean:137]:
```suggestion
theorem rootsExpandEquivRoots_apply (x) : (rootsExpandEquivRoots p f x : R) = x ^ p := rfl
```
Should this be a simp lemma?

File: Mathlib/FieldTheory/Perfect.lean [hunk 2]
```diff
@@ -111,14 +111,50 @@ theorem not_irreducible_expand (f : R[X]) : ¬ Irreducible (expand R p f) := by
...
+a bijection from the set of roots of `Polynomial.expand R (p ^ n) f` to the set of roots of `f`.
+It's given by `x ↦ x ^ (p ^ n)`, see `rootsExpandPowEquivRoots_apply`. -/
+noncomputable def rootsExpandPowEquivRoots :
+    (n : ℕ) → (expand R (p ^ n) f).roots.toFinset ≃ f.roots.toFinset
+  | 0 => Equiv.Set.ofEq <| by rw [pow_zero, expand_one]
+  | n + 1 => (Equiv.Set.ofEq <| by rw [pow_succ, ← expand_expand]).trans
+    (rootsExpandEquivRoots p (expand R (p ^ n) f)) |>.trans (rootsExpandPowEquivRoots n)
+
+theorem rootsExpandPowEquivRoots_apply (n : ℕ) (x) :
+    (rootsExpandPowEquivRoots p f n x).1 = x.1 ^ p ^ n := by
```
Comment (erdOne) [inline Mathlib/FieldTheory/Perfect.lean:149]:
```suggestion
    (rootsExpandPowEquivRoots p f n x : R) = x ^ p ^ n := by
```
Should this be one as well?

File: Mathlib/FieldTheory/Perfect.lean (hunk 1 above)
Comment (acmepjz) [reply Mathlib/FieldTheory/Perfect.lean:137]: > Should this be a simp lemma?

I'm not sure. As a reference, [frobenius_def](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/CharP/Basic.html#frobenius_def) and [iterate_frobenius](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Algebra/CharP/Basic.html#iterate_frobenius) are not simp lemmas.

Comment (erdOne) [reply Mathlib/FieldTheory/Perfect.lean]: Yes but this is the coercion of the application. After some more consideration, I think we should make this a simp lemma.

**Comment by erdOne** [comment]
maintainer merge

**Review by jcommelin — APPROVED** [review]
Thanks :tada:

bors merge

**Pull request closed without merging**